docs/contributing/component-js-configuration-options.md

```diff
@@ -0,0 +1,192 @@
+
+The order in which variables are written defines their priority, with objects passed sooner being overwritten by those passed later. As we want the user's configuration to take precedence over our defaults, we list our default configuration object first.
+
+There is no guarantee `config` will have any value at all, so it'll be `undefined`. We use an OR operator (`||`) as a safety check. If the value is `undefined`, we use an empty object instead.
+
+```javascript
+import { mergeConfigs } from '../../common/index.mjs'
+
+this.config = mergeConfigs(
+  defaultConfig,
+  config || {}
+)
+```
+
+We can now reference the configuration option anywhere we need to in the component's JavaScript:
+
+```javascript
+this.config.rememberExpanded
+// => true
+```
+
+It's now possible to individually initialise the component with configuration options.
+
+```html
+<script>
+  var $accordion = document.getElementById("accordion")
+  new window.GOVUKFrontend.Accordion($accordion, {
+    rememberExpanded: false
+  }).init()
+</script>
+```
+
+## Allowing options to be passed through the `initAll` function
+
+Usually, teams will not be individually initialising components. Instead, GOV.UK Frontend ships with an `initAll` function, which searches the page for instances of components and automatically initialises them.
+
+In `src/govuk/all.mjs`, update the component's `new` function to pass through a nested configuration object. The nested object should use the component's name converted to camelCase (for example, the 'Character Count' component becomes `characterCount`).
+
+```javascript
+new Accordion($accordion, config.accordion).init()
+```
+
+It's now possible to pass configuration options for your component, as well as multiple other components, using the `initAll` function.
+
+```html
+<script>
+  window.GOVUKFrontend.initAll({
+    accordion: {
+      rememberExpanded: false
+    },
+    characterCount: {
+      threshold: 90
+    }
+  })
+</script>
+```
+
+## Adding support for HTML `data-*` attributes
+
+For convenience, we also allow for configuration options to be passed through HTML `data-*` attributes.
+
+You can find `data-*` attributes in JavaScript by looking at an element's `dataset` property. Browsers will convert the attribute names from HTML's kebab-case to more JavaScript-friendly camelCase when working with `dataset`. See ['Naming configuration options'](#naming-configuration-options) for exceptions.
+
+As we expect configuration-related `data-*` attributes to always be on the component's root element (the same element with the `data-module` attribute), we can access them all using `$module.dataset`.
+
+Using the `mergeConfigs` call discussed earlier in this document, update it to include `$module.dataset` as the highest priority.
+
+```javascript
+import { mergeConfigs } from '../../common/index.mjs'
+import { normaliseDataset } from '../../common/normalise-dataset.mjs'
+
+this.config = mergeConfigs(
```

Comment:
Until we can use `Object.assign()` (or object spread) we should suggest types here 😩 

For example in the Character Count component, linting errors appear for:

```js
this.config.maxlength
this.config.threshold
```

Because without types, we appear to do calculations on unknown types

```suggestion
/** @type {CharacterCountConfig} */
this.config = mergeConfigs(
```

Reply:
Hard isn't it

Also the realisation that `normaliseDataset()` is a bit magic and will automatically picks types for you based on what the string value looks like. The compiler can't automatically "type" the return value.

Could potentially be solved by passing the defaults into `normaliseDataset()`

```js
this.config = normaliseDataset($module.dataset, {
  ...defaultConfig,
  ...config || {},
})
```

So rather than guessing types you could grab them from defaults

Not one for this PR, so yeah let's leave it